### ISSUE TITLE
Programmatic access to OpenMM constants 

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#1149]
Sometimes, it's useful to have programmatic access to physical constants used internally in OpenMM, such as when one needs to implement [a `CustomNonbondedForce` version of the Coulomb potential for alchemical modifications](https://github.com/choderalab/alchemy/blob/master/alchemy/alchemy.py#L544) that uses `ONE_4PI_EPS0` (currently defined in [`platforms/reference/include/SimTKOpenMMRealType.h`](https://github.com/pandegroup/openmm/blob/master/platforms/reference/include/SimTKOpenMMRealType.h#L145):

```
electrostatics_energy_expression += "U_electrostatics = ONE_4PI_EPS0*lambda_electrostatics*chargeprod/reff_electrostatics;"
```

Would there be some way to expose the same value of the constants used in OpenMM through a programmatic interface, such as through `simtk.unit`?

**Comment by peastman** [comment]
simtk.unit.constants looks like the place to add things like that.

**Comment by jchodera** [comment]
Are all the OpenMM constants collected in that header file?  Are they also
the same for all platforms?

If so, we can compile a list of useful constants from there, and add a
comment to update simtk.unit if they ever change.

**Comment by peastman** [comment]
I don't know what you mean by "the OpenMM constants".  There are certain constants that appear in that file.  All of them are physical constants, not code related things: Boltzmann's constant, the speed of light, etc.  The permittivity of free space (eps0) would be another very reasonable one to include.

**Comment by jchodera** [comment]
To clarify: There are a number of physical constants that OpenMM uses to compute energies.  Some are known to higher precision than the numbers printed in that file.  At some point, a decision was made to truncate these physical constants to some finite number of sig figs to represent in this file, or to use a consistent set of "compatible" physical constant values to compute energies.  For very large systems, knowing the exact constants employed by OpenMM is important to minimizing deviations from OpenMM energies when computing energies using Custom*Force classes.

The request: we should make any finite-precision truncated forms of physical constants used by OpenMM "under the hood" to compute energies available in order to allow users to compute energies using the same constants if needed.

**Comment by maciejj** [comment]
I agree with @jchodera 
For example the `ONE_4PI_EPS0` differs a little bit between Amber, CHARMM or NAMD simulation engines. What is more, the one used in OpenMM is not equal to any of them. The closest is the NAMD value. 

This cause problems with a reproducibility of the results, when comparing to other software. 
For example Amber charges are parametrized to be used with the Amber value of `ONE_4PI_EPS0`. I'm currently comparing OpenMM with AmberFF vs results from Amber using:
https://github.com/peastman/OpenMMEnergyComparisons
Scaling charges using NonbondedForce class or recompiling OpenMM with slightly changed value of the `ONE_4PI_EPS0` makes results better.

TL;DR
The values of the constants should be accessible via the API without need to recompile the OpenMM, because some forcefields were parametrized to be used with the specific values.

**Comment by swails** [comment]
Discussed extensively in ParmEd/ParmEd#426

**Comment by swails** [comment]
Also note that you can *effectively* change the electrostatic constant by scaling the charges by the ratio of the constants used in each program.  So maybe just publishing in the documentation exactly what the constant *is* is enough here.

**Comment by swails** [comment]
Oops, square root of the ratio :smiley:

**Comment by maciejj** [comment]
> scaling the charges by the ratio of the constants used in each program
This is exactly what I'm doing right now. 
It's working well for the case without cutoff and PBC. 
I'm also trying to find a way to get matching results when using cutoffs and PME.

**Comment by swails** [comment]
There's a lot more to do with PME -- sander uses 4th order splines, whereas OpenMM uses 5th order.  The default PME parameters are also different.

It's probably not impossible, but may very well be impossible with the sander Python API (since I don't know if I added the ability to change the PME order).

**Comment by peastman** [comment]
For reference (just to have the information all in one place), the value used in OpenMM is 138.935456 kJ/nm = 332.063709 kcal/angstrom.  That's basically identical to the LAMMPS value given in the ParmEd discussion

**Comment by maciejj** [comment]
I have also an issue when comparing VDW between Amber and Openmm in PBC system with cutoff. I think this is connected with the _long range dispersion correction_ which by default is on in Amber, but Off in Openmm.

**Comment by peastman** [comment]
If you use NonbondedForce, it's on by default.  In CustomNonbondedForce it's off by default, since it doesn't make sense with some types of nonbonded interactions.

**Comment by maciejj** [comment]
Thanks for the answer, from the http://docs.openmm.org/7.0.0/userguide/theory.html#nonbondedforce I thought It's off by default
> When using periodic boundary conditions, NonbondedForce can optionally add a term (known as a long range dispersion correction) to the energy that approximately represents the contribution from all interactions beyond the cutoff distance:

**Comment by swails** [comment]
> I have also an issue when comparing VDW energy between Amber and Openmm in PBC system with cutoff. I think this is connected with the long range dispersion correction which by default is on in Amber, but Off in Openmm.

Amber and OpenMM do things a little differently.  A few comments here:

1. Make sure you are using some kind of double precision platform (Reference or either GPU platform) when doing these comparisons
1. Amber and OpenMM compute the long-range correction *slightly* differently.  The principle behind the correction is the same, but Amber ignores the repulsive part of the potential in the long-range correction, whereas OpenMM does not.

In particular, the second point above is slightly more pronounced than I had expected (I think it was on the order of ~0.6 kcal/mol for a ~21K atom system with an 8 A cutoff -- it will be less for a larger cutoff), but will have no measurable effect on the simulation (after all, only differences matter, and most of that contribution is systematic and will cancel).  Turning off the long-range correction improves the agreement to <0.01 kcal/mol.  I actually figured this out for one of the first contributions I ever made to OpenMM in #351 -- [see the description here](https://github.com/pandegroup/openmm/issues/351#issue-28591611).

But that is enough to make Amber and OpenMM disagree on vdW terms more in PBC calculations than infinite cutoff, non-PBC calculations.

**Comment by peastman** [comment]
All physical constants used in C++ code should now be identical to the ones in simtk.unit.constants.
